pandas/core/indexing.py

```diff
@@ -961,12 +961,17 @@ class _LocIndexer(_LocationIndexer):
 
     @doc(_LocationIndexer._validate_key)
     def _validate_key(self, key, axis: int):
-
         # valid for a collection of labels (we check their presence later)
         # slice of labels (where start-end in labels)
         # slice of integers (only if in the labels)
-        # boolean
-        pass
+        # boolean not in slice and with boolean index
+        if not is_bool_dtype(self.obj.index) and isinstance(key, bool):
+            raise TypeError("Boolean label can not be used without a boolean index")
```

Comment:
Why not KeyError?

could you run asv? Iirc is_bool_dtype is not completly cheap, but might be wrong

Reply:
```
       before           after         ratio
     [1367cacd]       [ee07ee02]
                      <fix-20432>
+         323±7μs         553±70μs     1.71  indexing.AssignTimeseriesIndex.time_frame_assign_timeseries_index
+        52.1±1μs       71.0±0.4μs     1.36  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
+      38.4±0.9μs       49.4±0.8μs     1.29  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
+      19.6±0.6μs       24.0±0.2μs     1.22  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+     6.82±0.09ms       8.18±0.9ms     1.20  indexing.NumericSeriesIndexing.time_getitem_lists(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
+      49.8±0.7μs       59.7±0.3μs     1.20  indexing.DataFrameNumericIndexing.time_loc
+      20.3±0.5μs       23.8±0.6μs     1.18  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
+        66.0±1μs         76.8±2μs     1.16  indexing.NumericSeriesIndexing.time_loc_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+        91.9±2μs        106±0.8μs     1.16  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+        55.2±2μs         61.7±1μs     1.12  indexing.NumericSeriesIndexing.time_getitem_scalar(<class 'pandas.core.indexes.numeric.UInt64Index'>, 'unique_monotonic_inc')
+         104±2μs          115±2μs     1.11  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
-      2.27±0.3ms      1.97±0.02ms     0.87  indexing.IntervalIndexing.time_getitem_scalar
-        588±10μs          493±7μs     0.84  indexing.NumericSeriesIndexing.time_getitem_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'nonunique_monotonic_inc')
-       334±100μs          266±5μs     0.80  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Float64Engine'>, <class 'numpy.float64'>), 'monotonic_decr')
-       320±100μs          228±3μs     0.71  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Int32Engine'>, <class 'numpy.int32'>), 'monotonic_decr')
-        211±40μs          151±1μs     0.71  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Float64Index'>, 'nonunique_monotonic_inc')
-     1.98±0.06ms      1.33±0.05ms     0.67  indexing.NumericSeriesIndexing.time_getitem_array(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
```
That is a pretty big penalty indeed. What I could do is swap the statements so `is_bool_dtype` will only be called for boolean keys. In that case the performance reduction should only occur for boolean keys. Or is there a faster way to do this check? As far as I know there is no boolean index type? 

Regarding why I chose `TypeError` over `KeyError`: I think it depends if you think the issue is the key or the index? I have no objection to changing it. But at least for boolean slices, TypeError seems to be the correct error type.

Reply:
Hm we are using mostly KeyError, so I would prefer this here. 

Yeah checking the key first should help. I do not know a faster way

Reply:
Changed to KeyError and switched the boolean statements.

Reply:
Could you run asvs again?

Reply:
```
       before           after         ratio
     [cc84a238]       [b90967d7]
     <fix-40014~1>       <fix-20432>
+         113±3μs         203±60μs     1.79  indexing.DataFrameStringIndexing.time_boolean_rows
+      3.80±0.1μs       4.71±0.4μs     1.24  indexing.DataFrameStringIndexing.time_getitem_scalar
+     2.99±0.06ms      3.34±0.04ms     1.12  indexing.NumericSeriesIndexing.time_loc_array(<class 'pandas.core.indexes.numeric.Float64Index'>, 'unique_monotonic_inc')
+      42.9±0.3μs         47.3±1μs     1.10  indexing.NumericSeriesIndexing.time_iloc_list_like(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-      46.0±0.8μs       41.3±0.3μs     0.90  indexing.NumericSeriesIndexing.time_loc_slice(<class 'pandas.core.indexes.numeric.Int64Index'>, 'unique_monotonic_inc')
-        277±10μs          248±5μs     0.89  indexing_engines.NumericEngineIndexing.time_get_loc((<class 'pandas._libs.index.Float32Engine'>, <class 'numpy.float32'>), 'non_monotonic')
-        453±90μs          392±1μs     0.86  indexing.Take.time_take('int')
-        3.93±2μs      2.94±0.05μs     0.75  indexing_engines.ObjectEngineIndexing.time_get_loc('monotonic_incr')
```